Fix implemenation of snapshotAtTime for SnapshotDelta

TheodorRene opened the pull request:
Currenlty a timebased heurestic for finding the least expensive
operations, but it should be number-of-operations in the future

La også til en metode for activatedEntities som vi har sagt vi skal implmentere I overLeaf

Edit: Denne PRen ble helt utrolig kaotisk, beklager det. Rydder den opp på mandag. Men ser gjerne etter tilbakemelding på commitsene

TheodorRene committed:
Fix implementation of snapshotAtTime for SnapshotDelta

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -329,6 +336,14 @@ object SnapshotDeltaObject {
+   * Currently unused but might be relevant later
+   *
+   * @param instant   the timestamp
+   * @param snapshot1 snapshot1
+   * @param snapshot2 snapshot 2
```
```suggestion
   * @param snapshot1 snapshot1
   * @param snapshot2 snapshot2
```
before my ocd kills me 😂

TheodorRene replied:
haha disse scaladockommentarene får dessverre veldig lite kjærlighet

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -46,19 +46,26 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
+    // But for it to be feasible each graph should have a sequence ID as well
+    // This is possible, I think, but not implemented
+    val closestGraph = graphs.reduceLeft((snap1, snap2) =>
+      if (Duration.between(snap2.instant, instant).isNegative) { // This means the next snapshot in the list is past the instant
```
```suggestion
      if (snap2.instant > instant) { 
```
Er dette det samme? I så fall blir kommentaren mindre nødvendig også

TheodorRene replied:
ja, det er mye bedre faktisk. Ble litt låst i hodet mitt til å bruke Duration.between

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -46,19 +46,26 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
-      logger.warn("Closest graph is in the future. Need to apply logs forwards")
+    } else if (instant < closestGraph.instant) { // We are asking an instant before the initial materialized snapshot
+      logger.warn("Closest graph in the future, and there is no one in the past.")
+      createGraph(logs.map(l => (l.timestamp, l)).filterByRange(Instant.MIN, instant).map(_._2))
```
```scala
def getLogsBetween(logs: RDD[LogTSV], t1:Instant, t2: Instant): RDD[LogTSV] = {
      logs.map(log => (log.timestamp, log)).filterByRange(t1, t2))
}
```
Kan vi ekstrahere denne loggfiltreringen til en egen funksjon? Vi gjør vel den flere steder også(?) 
Isolerer den bedre for en eventuell test og

eivindkopperud replied:
nja trekker det tilbake, det er jo veldig deklarativt sånn det er egentlig. man lager basically en eksakt wrapper rundt sparkfunksjonalitet if that makes sense

TheodorRene replied:
nja, vil egentlig si meg enig at det burde være en funksjon. Og det du foreslo hadde ikke vært helt riktig. Du mangler en `.map(_._2)` som jeg synes personlig er ganske stygt, men nødvending. Jeg skal ta en kikk på det

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -46,19 +46,26 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
       closestGraph.graph
-    } else if (closestGraph.instant.isBefore(instant)) {
-      logger.warn("Closest graph is in the future. Need to apply logs forwards")
+    } else if (instant < closestGraph.instant) { // We are asking an instant before the initial materialized snapshot
+      logger.warn("Closest graph in the future, and there is no one in the past.")
```
```suggestion
    } else if (instant < closestGraph.instant) {
      logger.warn("There exists no snapshot before the given instant.")
```
Kommentar og logging sier litt det samme - er det dumt om vi bare beholder én?

TheodorRene replied:
fixed

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -46,19 +46,26 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
+    //Closest graph should be based on number of logs, not time.
+    // But for it to be feasible each graph should have a sequence ID as well
+    // This is possible, I think, but not implemented
+    val closestGraph = graphs.reduceLeft((snap1, snap2) =>
+      if (snap2.instant > instant) { 
+        snap1
+      } else {
+        snap2
+      })
```
Kan vi ekstrahere denne reducing-a til closestGraph-funksjonen? Kan hende jeg misforstår intensjonen her tho

eivindkopperud replied:
```suggestion
    val closestGraph = getClosestGraph(graphs, instant) 
```

```scala
def getClosestGraph(graph: SomeGraphType, instant: Instant): Snapshot =
    graphs.reduceLeft((snap1, snap2) =>
      if (snap2.instant > instant) { 
        snap1
      } else {
        snap2
      })
```

Eller er det kanskje ikke så rett frem kanskje, idk

TheodorRene replied:
jo den er ikke dum

eivindkopperud reviewed (approved):
Good stuff! Har lagt igjen noen kommentarer, men burde være good as is også! 👏

TheodorRene committed:
Add activatedEntities method to SnapshotDelta

Refactoring from PR #20 feedback

eivindkopperud commented on data-generation/src/main/scala/thesis/Landy.scala:
```diff
@@ -1,15 +1,16 @@
 package thesis
 
-import org.apache.spark.graphx.{Edge, EdgeRDD, EdgeTriplet, Graph, VertexRDD}
+import org.apache.spark.graphx._
```
Hvorfor importere alt?

TheodorRene replied:
Ingen peiling, det er Intellij/scalafmt som håndterer imports. Skriver dem nesten aldri manuelt

eivindkopperud replied:
Ahh, right. Er en Intellij-innstilling.  Skjønner egentlig ikke at det er default - wildcards er jo ikke noe bra.
Kan endres på Settings ->Editor -> Code Style -> Scala
![image](https://user-images.githubusercontent.com/43027083/165734401-88de4795-722a-455d-b66f-bd9ec7017f28.png)
Endre fra 5->100 elns

TheodorRene replied:
ah, tenkte det måtte være no sånt. Skal endre på det

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -17,16 +17,27 @@ import scala.util.Try
 
 
 //Should just be a trait
+// TODO refactor this into its own file, snapshotDelta is a very big file ATM
```
+1, digg

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -17,16 +17,27 @@ import scala.util.Try
 
 case class SnapshotEdgePayload(id: EdgeId, attributes: Attributes)
 
+//TODO have file with all the finurlige case classes
+case class Interval(start: Instant, stop: Instant)
```
+1 :finurlig:

eivindkopperud commented on data-generation/src/main/scala/thesis/SnapshotDelta.scala:
```diff
@@ -42,7 +53,13 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
   }
 
-  def backwardsApplyLogs(g: AttributeGraph, logsToApply: RDD[LogTSV]): AttributeGraph = throw new NotImplementedError()
+  override def activatedEntities(interval: Interval): (RDD[VertexId], RDD[EdgeId]) = {
+    val relevantLogs = getLogsInInterval(logs, interval)
+    val (vertexSquash, edgeSquash) = getSquashedActionsByEntityId(relevantLogs)
+    val activatedVertices = vertexSquash.filter(_._2.action == CREATE).map(_._1)
+    val activatedEdges = edgeSquash.filter(_._2.action == CREATE).map(_._1)
+    (activatedVertices, activatedEdges)
+  }
```
Nice, men jeg tror vi gjør oss en tjeneste om vi reduserer omfanget til pull requestene vi lager.

TheodorRene replied:
Ja, støtter den. Var bare i så god flyt på denne branchen at jeg ikke orket å lage en ny

TheodorRene replied:
Jeg la til enda mer kaos

TheodorRene committed:
Refactor types(case classes) into a single file

The reasoning is that you often just use go to definition to find these datatypes

TheodorRene commented on data-generation/src/main/scala/factories/LandyGraphFactory.scala:
```diff
@@ -40,7 +42,7 @@ object LandyGraphFactory {
           validFrom = t2,
           validTo = t3,
           attributes = HashMap(
-            "color"->"orange",
```
hmm, du som ikke kjører formatteren på denne filen, eller min formatter som har andre regler?

eivindkopperud replied:
Hehe 😅 Jeg kan begynne å kjøre formattereren jeg 😂

TheodorRene replied:
var du som fikk meg til å begynne med format-on-save! haha

eivindkopperud replied:
👀 👀 
im such a hypocrite

eivindkopperud reviewed (approved):
Synes denne ser bra ut nå! Vi kunne lagd tester, men jeg tror egt det går bra as is, og heller bare ha det i bakhodet dersom vi har rar oppførsel. Viktigste er at vi gjør det testbart tenker jeg!

bare å smelle inn denne 🚀
